File: java/perfect-numbers/src/main/java/NaturalNumber.java
import java.util.stream.IntStream;

class NaturalNumber {

  private Classification classification;

  NaturalNumber(int n) {
    if (n < 1) {
      throw new IllegalArgumentException("You must supply a natural number (positive integer)");
    }

    int sum = IntStream.rangeClosed(1, n / 2).filter(i -> n % i == 0).sum();

    if (sum == 1) {
      this.classification = Classification.DEFICIENT;
    } else if (sum == n) {
      this.classification = Classification.PERFECT;
    } else if (sum > n) {
      this.classification = Classification.ABUNDANT;
    } else {
      this.classification = Classification.DEFICIENT;
    }
  }

  public Classification getClassification() {
    return this.classification;
  }

}

File: kotlin/perfect-numbers/src/main/kotlin/NaturalNumber.kt
enum class Classification {
    DEFICIENT, PERFECT, ABUNDANT
}

fun classify(naturalNumber: Int): Classification {
    if (naturalNumber < 1) {
        throw RuntimeException("Not a natural number")
    }

    val sum = (1..naturalNumber / 2).map {
        if (naturalNumber % it == 0) it else 0
    }.sum()

    return when {
        sum == naturalNumber -> Classification.PERFECT
        sum < naturalNumber -> Classification.DEFICIENT
        else -> Classification.ABUNDANT
    }
}
